test/generate_test_advanced.c

```diff
@@ -0,0 +1,1117 @@
+      int i;
+      bool changes = true;
+      while (changes){
+        changes = false;
+        for (i = 0; i < guard_stmt->n; ++i) {
+          changes = changes || update_bounds_with_equation(param_bounds,
+              &guard_stmt->eq[i], margin);
+        }
+        modified = modified || changes;
+      }
+      modified = modified || look_for_bounds_in_ast(guard_stmt->then, param_bounds, margin);
+    }
+    else if (CLAST_STMT_IS_A(stmt, stmt_block)) {
+      struct clast_block *block_stmt = (struct clast_block*) stmt;
+      modified = look_for_bounds_in_ast(block_stmt->body, param_bounds, margin);
+    }
+  }
+  return modified;
+}
+
+static void print_statement_macro(FILE *out, struct bounds *bounds){
+  unsigned i;
+
+  fprintf(out, "#define S1(");
+  for (i = 0; i < bounds->names.nb_names; ++i) {
+    if(i)
+      fprintf(out, ", ");
+    fprintf(out, "p%u", i);
+  }
+  fprintf(out, ") do {");
+  fprintf(out, "h = h_good;\\\ngood(");
+  for (i = 0; i < bounds->names.nb_names; ++i) {
+    if(i != 0)
+      fprintf(out, ", ");
+    fprintf(out, "p%u", i);
+  }
+  fprintf(out, ");\\\nh_good = h;\\\n");
+  fprintf(out, "h = h_test;\\\ntest(");
+  for (i = 0; i < bounds->names.nb_names; ++i) {
+    if(i != 0)
+      fprintf(out, ", ");
+    fprintf(out, "p%u", i);
+  }
+  fprintf(out, ");\\\nh_test = h;\\\n");
+  fprintf(out, "if ( h_good != h_test ) {\\\n");
+  fprintf(out,
+      "fprintf(stderr, \"\\x1b[1m\\x1b[5m\\x1b[31mTest failed\\x1b[0m\\n\");\\\n");
+  fprintf(out,
+      "fprintf(stderr, \"\\x1b[1m\\x1b[31mThis happened with the following parameters:\\n\");\\\n");
+  fprintf(out, "fprintf(stderr, \"(");
+  for (i = 0; i < bounds->names.nb_names; ++i) {
+    if(i != 0)
+      fprintf(out, ", ");
+    fprintf(out, "%%d");
+  }
+  fprintf(out, ")\\n\"");
+  for (i = 0; i < bounds->names.nb_names; ++i) {
+    fprintf(out, ", ");
+    fprintf(out, "p%u", i);
+  }
+  fprintf(out, ");\\\n");
+  fprintf(out, "exit(EXIT_FAILURE);\\\n} } while(0)\n");
+}
+
+static void print_good_test_declaration(FILE *out, struct bounds *bounds){
+  unsigned int i;
+  fprintf(out, "void test(");
+  for (i = 0; i < bounds->names.nb_names; ++i) {
+    if(i)
+      fprintf(out, ", ");
+    fprintf(out, "int p%u", i);
+  }
+  fprintf(out, ");\n");
+  fprintf(out, "void good(");
+  for (i = 0; i < bounds->names.nb_names; ++i) {
+    if(i)
+      fprintf(out, ", ");
+    fprintf(out, "int p%u", i);
+  }
+  fprintf(out, ");\n\n");
+}
+
+static void fprint_cloog_program_parameters_decl(FILE *out, CloogProgram *p){
+  int i;
+  for (i = 0; i < p->names->nb_iterators; ++i){
+    fprintf(out, "int %s;\n", p->names->iterators[i]);
+  }
+}
+
+static const char initial_hash_value[] = "2166136261u";
+
+static const char preamble1[] =
+"#include <stdio.h>\n"
+"#include <stdlib.h>\n"
+"\n"
+"static unsigned h;\n"
+"\n"
+"void hash(int v)\n"
+"{\n"
+"  size_t i;\n"
+"  union u {\n"
+"    int v;\n"
+"    unsigned char c[sizeof(int)];\n"
```

Comment:
Could you do the change from c[1] to c[sizeof(int)] in generate_test.c as well so the commit has all desired updates ? Thanks.